### PR TITLE
Multimethods for the Linear Algebra module

### PR DESCRIPTION
**Pull request by joaosferreira**
I've picked up where @prasunanand left off in https://github.com/Quansight-Labs/unumpy/pull/48 by adding additional multimethods + tests for the Linear Algebra module (linalg). These include:
- linalg.multi_dot
- linalg.matrix_power
- linalg.solve
- linalg.tensorsolve
- linalg.lstsq
- linalg.inv
- linalg.pinv
- linalg.tensorinv

**Note:** Adding some default implementations may be relevant here.

### CODE REVIEW

File: unumpy/linalg/_multimethods.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
+import functools
+import operator
+from uarray import create_multimethod, mark_as, all_of_type, Dispatchable
+import builtins
+
+create_numpy = functools.partial(create_multimethod, domain="numpy")
```
Comment (hameerabbasi) [inline unumpy/linalg/_multimethods.py]:
```suggestion
create_numpy = functools.partial(create_multimethod, domain="numpy.linalg")
```

File: unumpy/numpy_backend.py [hunk 2]
```diff
@@ -24,11 +24,13 @@ def __ua_function__(method, args, kwargs):
     if method in _implementations:
         return _implementations[method](*args, **kwargs)
 
-    if not hasattr(np, method.__name__):
+    if hasattr(np, method.__name__):
+        return getattr(np, method.__name__)(*args, **kwargs)
+    elif hasattr(np.linalg, method.__name__):
+        return getattr(np.linalg, method.__name__)(*args, **kwargs)
+    else:
```
Comment (hameerabbasi) [inline unumpy/numpy_backend.py]: I'd do this dispatching based on `method.domain`. Something like
```python
module = np
domain_hierarchy = method.domain.split('.')
for d in domain_hierarchy[1:]:
    module = getattr(module, d)

implementation = getattr(module, method.__name__)
return implementation(*args, **kwargs)
```

**Review by hameerabbasi — CHANGES_REQUESTED** [review]
A few small comments to make everything a bit more general.

**Comment by hameerabbasi** [comment]
I'd add defaults to those that are simple to do here. For the more complicated ones, leave them out.

**Comment by joaosferreira** [comment]
I'll keep adding more.

File: unumpy/linalg/_multimethods.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+
+@create_numpy(_first_argreplacer)
+@all_of_type(ndarray)
+def multi_dot(arrays):
+    return arrays
+
+
+def matrix_power_default(a, n):
+    if a.shape[0] == a.shape[1]:
```
Comment (hameerabbasi) [inline unumpy/linalg/_multimethods.py]: This needs to be re-written in terms of the eigenvalue decomposition. If you're unsure what that is, here's a Wikipedia page (in NumPy it's the `eig` function): https://en.wikipedia.org/wiki/Eigendecomposition_of_a_matrix

https://math.stackexchange.com/a/2552148/59208

Comment (joaosferreira) [reply unumpy/linalg/_multimethods.py]: You are right, I misunderstood what matrix_power does when I wrote that.

File: unumpy/_multimethods.py [hunk 4]
```diff
@@ -1156,3 +1156,21 @@ def _allclose_default(a, b, rtol=1e-05, atol=1e-08, equal_nan=False):
 @all_of_type(ndarray)
 def allclose(a, b, rtol=1e-05, atol=1e-08, equal_nan=False):
     return a, b
+
+
+def diag_default(v, k=0):
+    if ndim(v) == 1:
+        a = zeros((len(v), len(v)), dtype=v.dtype)
+        for i in range(len(v)):
+            a[i, i] = v[i]
```
Comment (joaosferreira) [inline unumpy/_multimethods.py]: I haven't included the test for `diag`. The Sparse backend fails in this line with the following error:

> TypeError: 'COO' object does not support item assignment

Comment (hameerabbasi) [reply unumpy/_multimethods.py]: I would avoid defaults with long loops. :)

File: unumpy/linalg/_multimethods.py [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+        norm = x
+    elif ndim(x) == 1:
+        if ord is None:
+            norm = sqrt(sum(square(x)))
+        elif ord == inf:
+            norm = max(x)
+        elif ord == NINF:
+            norm = min(x)
+        elif ord == 0:
+            norm = sum(x[x != 0])
```
Comment (peterbell10) [inline unumpy/linalg/_multimethods.py]: What does the `x[x != 0]` term do? If `x == 0` then it wouldn't contribute to the sum anyway.

Looking at the numpy docs, it says `sum(x != 0)` which would be the same as `count_nonzero(x)`.

File: unumpy/linalg/_multimethods.py [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+    return (a,)
+
+
+@create_numpy(_self_argreplacer)
+@all_of_type(ndarray)
+def eigvalsh(a, UPLO="L"):
+    return (a,)
+
+
+def norm_default(x, ord=None, axis=None, keepdims=False):
```
Comment (peterbell10) [inline unumpy/linalg/_multimethods.py]: The axis argument is ignored

File: unumpy/_multimethods.py (hunk 4 above)
Comment (joaosferreira) [reply unumpy/_multimethods.py]: How can we return the diagonal matrix without using a loop and item assignment then?

File: unumpy/linalg/_multimethods.py (hunk 5 above)
Comment (joaosferreira) [reply unumpy/linalg/_multimethods.py]: I though it was the sum of all non-zero elements which is what it does but you are right, it's a count and not a sum.

File: unumpy/linalg/_multimethods.py (hunk 6 above)
Comment (joaosferreira) [reply unumpy/linalg/_multimethods.py]: Yes, some norms for matrices (nuclear norm, 2-norm and smallest singular value) are also ignored, I will try to add them as well if they are simple enough.

I'm still confused about how to use the axis argument but I will take a closer look. Can I assume only two-dimensional arrays are given? Because these might be easier for me.

Comment (peterbell10) [reply unumpy/linalg/_multimethods.py]: From the docs:

>If axis is an integer, it specifies the axis of x along which to compute the vector norms. If axis is a 2-tuple, it specifies the axes that hold 2-D matrices, and the matrix norms of these matrices are computed.

You've implemented the `axis=None` case. But if `isinstance(axis, int)` then you should do a vector norm over that axis,regardless of `ndim(x)`. That should be as simple as calling the underlying reduction operation (`sum`, `min` or `max`) with the `axis` and `keepdims` arguments.

You can also have `axis` be a pair `(i, j)` which gives the dimensions for a matrix norm. That would mean where you've used `axis=0` here, you should instead use `axis=axis[0]` and similarly `axis=1` becomes `axis=axis[1]`.

File: unumpy/_multimethods.py (hunk 4 above)
Comment (hameerabbasi) [reply unumpy/_multimethods.py]: In general, we'd avoid even having defaults with long loops or assignment. We'd just not have a default in that case.

**Comment by hameerabbasi** [comment]
A note: Correctness is better than completeness here. If a kwarg isn't supported, raise a `BackendNotImplementedError`.

File: unumpy/_multimethods.py (hunk 4 above)
Comment (dhirschfeld) [reply unumpy/_multimethods.py]: Not sure it's relevant in this case, but the diagonal is just a strided view of the flattened array:
```python
In [7]: N = 5

In [8]: A = np.arange(N*N).reshape(N, N)

In [9]: np.diag(A)
Out[9]: array([ 0,  6, 12, 18, 24])

In [10]: A.ravel()[0::(N+1)]
Out[10]: array([ 0,  6, 12, 18, 24])

In [11]: A.ravel()[0::(N+1)] = 9

In [12]: A
Out[12]:
array([[ 9,  1,  2,  3,  4],
       [ 5,  9,  7,  8,  9],
       [10, 11,  9, 13, 14],
       [15, 16, 17,  9, 19],
       [20, 21, 22, 23,  9]])
```

File: unumpy/linalg/_multimethods.py (hunk 6 above)
Comment (joaosferreira) [reply unumpy/linalg/_multimethods.py]: This was very useful, thank you!

File: unumpy/_multimethods.py (hunk 4 above)
Comment (joaosferreira) [reply unumpy/_multimethods.py]: Yes, extremely relevant in fact. This helped me reimplement `diag_default`.

File: unumpy/_multimethods.py [hunk 7]
```diff
@@ -1154,3 +1156,23 @@ def _allclose_default(a, b, rtol=1e-05, atol=1e-08, equal_nan=False):
 @all_of_type(ndarray)
 def allclose(a, b, rtol=1e-05, atol=1e-08, equal_nan=False):
     return a, b
+
+
+def diag_default(v, k=0):
+    if ndim(v) == 1:
+        zeros = [0 for _ in range(len(v) * len(v))]
```
Comment (hameerabbasi) [inline unumpy/_multimethods.py]: You should use the `zeros` multimethod here.

File: unumpy/_multimethods.py [hunk 8]
```diff
@@ -1154,3 +1156,23 @@ def _allclose_default(a, b, rtol=1e-05, atol=1e-08, equal_nan=False):
...
+        zeros[0 :: len(v) + 1] = v
+        return array(zeros).reshape((len(v), len(v)))
+    elif ndim(v) == 2:
+        if k < 0:
+            return array([v[i + abs(k), i] for i in range(N - abs(k))])
```
Comment (hameerabbasi) [inline unumpy/_multimethods.py]: And indexing here.

File: unumpy/_multimethods.py [hunk 9]
```diff
@@ -1154,3 +1156,23 @@ def _allclose_default(a, b, rtol=1e-05, atol=1e-08, equal_nan=False):
...
+        else:
+            return array([v[i, i + k] for i in range(N - k)])
```
Comment (hameerabbasi) [inline unumpy/_multimethods.py]: And here.

File: unumpy/linalg/_multimethods.py [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+        raise ValueError("Array must be two-dimensional")
+
+    if not isinstance(n, int):
+        raise ValueError("Exponent must be an integer.")
+
+    if a.shape[0] == a.shape[1]:
+        if n == 0:
+            return eye(a.shape[0], a.shape[1])
+        elif n < 0:
+            a = inv(a)
```
Comment (hameerabbasi) [inline unumpy/linalg/_multimethods.py]: You don't need this code at all.

**Review by hameerabbasi — COMMENTED** [review]
Some comments. In general, avoid loops if at all possible.

File: unumpy/linalg/_multimethods.py (hunk 10 above)
Comment (joaosferreira) [reply unumpy/linalg/_multimethods.py]: Like this?
```python
def matrix_power_default(a, n):
    if n == 0:
        return eye(a.shape[0], a.shape[1])
    elif n < 0:
        a = inv(a)

    eigenvalues, eigenvectors = eig(a)

    diagonal = diag(eigenvalues)

    return multi_dot([eigenvectors, power(diagonal, abs(n)), inv(eigenvectors)])
```

Comment (hameerabbasi) [reply unumpy/linalg/_multimethods.py]: Like this:
```python
def matrix_power_default(a, n):
    eigenvalues, eigenvectors = eig(a)
    diagonal = diag(eigenvalues)
    return multi_dot([eigenvectors, power(diagonal, n), inv(eigenvectors)])
```

File: unumpy/_multimethods.py (hunk 8 above)
Comment (joaosferreira) [reply unumpy/_multimethods.py]: Not sure I understand what you mean by indexing here. Could you elaborate a bit more?

Comment (hameerabbasi) [reply unumpy/_multimethods.py]: TBH, I'd remove the default for `diag` outright. It requires either large loops (which we don't like) or mutability, which a lot of libraries don't have.

File: unumpy/_multimethods.py (hunk 7 above)
Comment (joaosferreira) [reply unumpy/_multimethods.py]: As I mentioned before, the Sparse backend fails when I try to assign the diagonal values after creating the array with `zeros` multimethod.

Should I change back to using `zeros` and assignment and add an exception in tests for this backend or just remove this default implementation? Sorry for the confusion.

Comment (hameerabbasi) [reply unumpy/_multimethods.py]: Either is fine by me. 😄 I prefer the former.

File: unumpy/_multimethods.py (hunk 8 above)
Comment (joaosferreira) [reply unumpy/_multimethods.py]: Okay, this answers the question I made seconds ago. Thank you!

File: unumpy/linalg/_multimethods.py (hunk 10 above)
Comment (hameerabbasi) [reply unumpy/linalg/_multimethods.py]: Or even this:
```python
def matrix_power_default(a, n):
    eigenvalues, eigenvectors = eig(a)
    return (eigenvectors * power(eigenvalues, n)) @ inv(eigenvectors)
```

Comment (joaosferreira) [reply unumpy/linalg/_multimethods.py]: Referencing the NumPy docs:

> For positive integers n, the power is computed by repeated matrix squarings and matrix multiplications. If `n == 0`, the identity matrix of the same shape as M is returned.
If `n < 0`, the inverse is computed and then raised to the `abs(n)`.

Does this implementation ignore the `n == 0` and `n < 0`  cases then?

Comment (hameerabbasi) [reply unumpy/linalg/_multimethods.py]: > Does this implementation ignore the n == 0 and n < 0 cases then?

No, it just uses the properties of matrices. 😄

First, that `diag(power(a, n)) == matrix_power(diag(a), n)` for `ndim(a) == 1`. That is, if `A` is diagonal, then raising it to a power means raising all its diagonal elements to that power.

Second, `X @ diag(a) == a * X`. That is, if diagonal `A` is multiplied by a matrix from the left, the rows are just multiplied by the corresponding diagonal element.

Comment (hameerabbasi) [reply unumpy/linalg/_multimethods.py]: (I edited the implementation a few times to get it right, but it's now okay)

Comment (hameerabbasi) [reply unumpy/linalg/_multimethods.py]: You can test it out yourself:

```python
>>> a = np.diag(np.arange(5))
>>> b = np.ones((5, 5))
>>> a @ b
array([[0., 0., 0., 0., 0.],
       [1., 1., 1., 1., 1.],
       [2., 2., 2., 2., 2.],
       [3., 3., 3., 3., 3.],
       [4., 4., 4., 4., 4.]])
>>> b @ a
array([[0., 1., 2., 3., 4.],
       [0., 1., 2., 3., 4.],
       [0., 1., 2., 3., 4.],
       [0., 1., 2., 3., 4.],
       [0., 1., 2., 3., 4.]])
>>> np.diag(a) * b
array([[0., 1., 2., 3., 4.],
       [0., 1., 2., 3., 4.],
       [0., 1., 2., 3., 4.],
       [0., 1., 2., 3., 4.],
       [0., 1., 2., 3., 4.]])
>>> np.diag(a).reshape((len(a), 1)) * b
array([[0., 0., 0., 0., 0.],
       [1., 1., 1., 1., 1.],
       [2., 2., 2., 2., 2.],
       [3., 3., 3., 3., 3.],
       [4., 4., 4., 4., 4.]])
```

File: unumpy/linalg/_multimethods.py [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+    _identity_argreplacer,
+    _self_argreplacer,
+    _dtype_argreplacer,
+    mark_dtype,
+    _first_argreplacer,
+    _first2argreplacer,
+)
+
+
+@create_numpy(_first_argreplacer)
```
Comment (hameerabbasi) [inline unumpy/linalg/_multimethods.py]: I'd add a default here, it's easy:

```
ret = arrays[0]
for a in arrays[1:]:
    ret = ret @ a

return ret
```

Comment (hameerabbasi) [reply unumpy/linalg/_multimethods.py]: One can do a more sophisticated algorithm depending on the shapes, but it's okay for now.

File: unumpy/linalg/_multimethods.py [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+        else:
+            raise ValueError("Invalid norm order for vectors.")
+    else:
+        if axis[0] is None:
+            axis = (0, 1)
+        elif axis[0] == axis[1]:
+            raise ValueError("Duplicate axes given.")
+
+        if ord is None or ord == "fro":
+            return sqrt(sum(square(x), keepdims=keepdims))
```
Comment (peterbell10) [inline unumpy/linalg/_multimethods.py]: This will need to be two sums over `axis[0]` then `axis[1]`.

File: unumpy/linalg/_multimethods.py [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+                sum(x, axis=axis[1], keepdims=keepdims), keepdims=keepdims
+            ).astype(float)
+        elif ord == 1:
+            return max(
+                sum(x, axis=axis[0], keepdims=keepdims), keepdims=keepdims
+            ).astype(float)
+        elif ord == -1:
+            return min(
+                sum(x, axis=axis[0], keepdims=keepdims), keepdims=keepdims
+            ).astype(float)
```
Comment (peterbell10) [inline unumpy/linalg/_multimethods.py]: The `min`s and `max`s here need to be over the other axis.

File: unumpy/linalg/_multimethods.py [hunk 14]
```diff
@@ -0,0 +1,224 @@
...
+
+
+@create_numpy(_self_argreplacer)
+@all_of_type(ndarray)
+def eigvalsh(a, UPLO="L"):
+    return (a,)
+
+
+def norm_default(x, ord=None, axis=None, keepdims=False):
+    if axis is None or isinstance(axis, int):
```
Comment (peterbell10) [inline unumpy/linalg/_multimethods.py]: I would move all of the `axis=None` handling to the top of the function here.

File: unumpy/linalg/_multimethods.py (hunk 11 above)
Comment (joaosferreira) [reply unumpy/linalg/_multimethods.py]: Could we write `multi_dot_default` with respect to `dot`?

A simple implementation of `dot`:
```python
def dot_default(a, b, out=None):
    if ndim(a) == 0 or ndim(b) == 0:
        return a * b

    return a @ b


@create_numpy(_first2argreplacer, default=dot_default)
@all_of_type(ndarray)
def dot(a, b, out=None):
    return (a, b, out)
```

And then rewrite `multi_dot_default` as:
```python
def multi_dot_default(arrays):
    res = arrays[0]
    for a in arrays[1:]:
        res = dot(res, a)

    return res
```
A more sophisticated handling of shapes could also be done in `dot`.

File: unumpy/linalg/__init__.py [hunk 15]
```diff
@@ -0,0 +1 @@
+from ._multimethods import *
```
Comment (hameerabbasi) [inline unumpy/linalg/__init__.py:1]: Add `__all__`.

Comment (joaosferreira) [reply unumpy/linalg/__init__.py:1]: I've added `__all__`. I wasn't sure if the import statement should stay or not.

**Comment by joaosferreira** [comment]
The default implementation for `norm` was removed and it will be used in a future PR for adding tests that compare its results to NumPy's `norm`. This is intended as a testbed for creating robust defaults by testing them on corner cases. I'll open an issue for this.

File: unumpy/linalg/__init__.py (hunk 15 above)
Comment (hameerabbasi) [reply unumpy/linalg/__init__.py:1]: The `import` will stay... Probably move `__all__` to `_multimethods.py`

File: unumpy/linalg/__init__.py [hunk 16]
```diff
@@ -1 +1,24 @@
 from ._multimethods import *
+
+__all__ = [
```
Comment (hameerabbasi) [inline unumpy/linalg/__init__.py]: Must be a list of strings, not a list of what's visible. E.g. the first element will be `"multi_dot"` vs `multi_dot`.

**Review by hameerabbasi — CHANGES_REQUESTED** [review]
A few small changes left.

**Review by hameerabbasi — APPROVED** [review]
Thanks, @joaosferreira!
